javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+             case (l, r, i) =>
+               (Node(Right(("", s(0) + "")), Left("")) +: l, r, i)
+           }
+         case '+' | '*' => // greedy or not greedy
+           if (s.length >= 2 && s(1) == '?') {
+             parse(s.substring(2), nextGroupIndex) match {
+               case (l, r, i) =>
+                 (Node(Right(("", s(0) + "" + s(1))), Left("")) +: l, r, i)
+             }
+           } else {
+             parse(s.substring(1), nextGroupIndex) match {
+               case (l, r, i) =>
+                 (Node(Right(("", s(0) + "")), Left("")) +: l, r, i)
+             }
+           }
+         case '{' => // parse until end of occurrence
+           val (inside, remaining) = parseUntilEndBrace(s.substring(1))
+           parse(remaining, nextGroupIndex) match {
+             case (l, r, i) =>
+                (Node(Right(("", "{" + inside)), Left("")) +: l, r, i)
+           }
+         case '[' =>
+            val (inside, remaining) = parseUntilEndSquareBracket(s.substring(1))
+           parse(remaining, nextGroupIndex) match {
+             case (l, r, i) =>
+                (Node(Right(("", "")), Left("[" + inside)) +: l, r, i)
+           }
+         case e => 
+           parse(s.substring(1), nextGroupIndex) match {
+             case (Node(Right(("","")), Left(regex)) +: tail, remaining, nextGroupIndex) =>
+               (Node(Right(("","")), Left(e + regex)) +: tail, remaining, nextGroupIndex)
+             case (l: Seq[Node], remaining, nextGroupIndex) =>
+               (Seq(Node(Right(("","")), Left("" + e))) ++ l, remaining, nextGroupIndex)
+           }
+         }
+         res match {
+           case (Node(Right((left, right)), m) +:
+                 Node(Right(("", modifier)), Left("")) +: tail, r, i)
+                 if modifier == "?" || modifier == "*" || modifier == "+" || modifier == "*?"
+                  || modifier == "+?" || modifier.startsWith("{") =>
+                  
+                (Node(Right((left, right + modifier)), m) +: tail, r, i)
+           case ((n@Node(Left(_), m)) +:
+                 Node(Right(("", modifier)), Left("")) +: tail, r, i)
+                 if modifier == "?" || modifier == "*" || modifier == "+" || modifier == "*?"
+                  || modifier == "+?" || modifier.startsWith("{") =>
+                 (Node(Right(("", modifier)), Right(Seq(n))) +: tail, r, i)
+           case _ => res
+         }
+       }
+       parse(pattern, 1)._1 match {
+         case Nil => Node(Right(("","")), Left(""))
+         case Seq(n:Node) => n
+         case s => Node(Right(("","")), Right(s))
+       }
+    }
+    
+    val node = parseRegex(regex)
+    node.setNewGroup(1)
+    val allMatchingRegex_s = node.buildRegex
+    val allMatchingRegex = new js.RegExp(allMatchingRegex_s)
+    
+    @inline def finish(): (Int => Int) = {
+      val last = ensureLastMatch(0).toOption match {
+        case None => return (i: Int) =>
+          throw new Exception("regex did not match")
```

Comment:
Avoid `return`. Instead, put the rest of the code in the `case Some`

Reply:
Done.